Feature: Causality

avisionh opened the pull request:
# Summary

This branch models the possible causes for wellbeing responses. This is to understand what variables are most influential in affecting a student's wellbeing in the survey.

The key scripts where the analysis takes place are:
1. `notebooks/causality/01_preprocess.R` - wrangles the data.
2. `notebooks/causality/02_postprocess.R` - wrangles the data some more.
3. `notebooks/causality/03_traintest.R` - splits data into train and test sets.
4. `notebooks/causality/041_model_mr.R` - applies a multiple linear regression on data.
5. `notebooks/causality/042_model_rf.R` - applies a random forest on the data.
6. `notebooks/causality/043_model_best.R` - picks the best model from regression and random forest; computes variable importance plots.
7. `notebooks/causality/051_model_mr_updownsampling.R` - does same as before, but accounts for class imbalances in our dependent variable, `mode_wellbeing`, by up-sampling the smaller classes and down-sampling the larger classes.
8. `notebooks/causality/051_model_rf_updownsampling.R` - does same as before, but accounts for class imbalances in our dependent variable, `mode_wellbeing`, by up-sampling the smaller classes and down-sampling the larger classes.
9. `notebooks/causality/051_model_updownsampling_best.R` - picks the best model from regression and random forest; computes variable importance plots.

> **Note:** *Most of the other scripts, specifically those below are rebased from https://github.com/ukgovdatascience/surveyanalysis/pull/4, so can be ignored in this script.*
> - `notebooks/longitudinal-tracking/analysis.R`
> - `src/make_data/link_data.R`
> - `src/utils/read_responses.R`

# Checklists

<!--
These are DO-CONFIRM checklists; it CONFIRMs that you have DOne each item.

Outstanding actions should be completed before reviewers are assigned; if actions are irrelevant, please try and add a
comment stating why.

Incomplete pull/merge requests MAY be blocked until actions are resolved, or closed at the reviewers' discretion.
-->

### General checks

- [ ] Code runs
- [ ] Developments are **secure** and [**ethical**](https://www.gov.uk/government/publications/data-ethics-framework)
- [ ] You have made proportionate checks that the code works correctly
- [ ] Test suite passes
- [ ] Assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if necessary
- [ ] [Minimum usable documentation](http://agilemodeling.com/essays/documentLate.htm) written in the `docs` folder

### Project-specific checks

> These are additional checks determined necessary according to the analytical quality assurance plan
(see `docs/aqa/aqa_plan.md`).

- [ ] The methodology is sound.
- [ ] Are no blatant code errors that will mess up the modelling.
- [ ] The inferences drawn are reasonable.

### References
The approach taken was blatantly copied from the [tidymodels](https://www.tidymodels.org/start/case-study/) framework.

avisionh committed:
Extract questions and context

This is so we can understand the predictive factors of wellbeing and learning scores.

avisionh committed:
Load data for analysis

This is so we can understand drivers behind wellbeing and learning.

avisionh committed:
Compute summary mode for each dependent variable

This is so we have a single target variable to aim for.

avisionh committed:
Add dependent variable

This is so we have a target variable to measure with.

avisionh committed:
Extract IV for modelling

This is so our data is suitable for modelling.

avisionh committed:
Tidy up folders and files

This is so they are organised correctly.

avisionh committed:
Combine IV and DV into same df

This is so we can model it. Also create a lag dependent variable to use as an independent variable.

avisionh committed:
Try multinomial regression on data

This is to see how well it performs on a train and validation set.

avisionh committed:
Try random forest on data

This is to see how well it performs on a train and validation set and compare with multinomial regression.

avisionh committed:
Fit last model

This is so we can see how well it performs against the test set.

The last model is the best one out of multinomial regression and random forest that we employed.

avisionh committed:
Use up- and down-sampling

This is so we can see if this improves our performance.

Also resturcture files to organise better.

avisionh committed:
Apply up-down-sampling for multinomial regression

This is so we can compare with the random forest.

avisionh committed:
Fit best model with up-down-sampling

This is so we can see if it is better than when we did not do up- and down-sampling.

avisionh committed:
Bump

martinctc commented:
Great work @avisionh! 

I've reviewed this PR and had the following observations:

1. Computing lag for IV is a nice detail to add - I like this!

2. Would it be worth running PCA or factor analysis on the predictors, if you haven't yet? Since multi-collinearity is a typical challenge with survey datasets. 

3. Agree with multinomial regression and RF as the model choices to try with this set of data.

4. When you run `vip()` for either models, do the models come up with similar results on the most important independent variables? It's usually difficult to get good accuracy on survey data, so I'd typically also do a sense check on whether the results make sense, based on domain knowledge / context. 

5. Also, how do the variable importance plots compare with correlation coefficients vs the dependent variable? It's usually helpful to be able to explain in _both_ cases of (i) either they are similar or (ii) very different. 

6. Presumably the data covers all the pupils in the analysis 'universe'? i.e. it's technically a census rather than a sample of the universe? If it's the latter, it could be worth thinking about weighting, but this might not be necessary if we think it's a fair representation of the population.

Note that I did not review the code line-by-line, and the notes are only based on a review of the overall methodology. The code however looks very neat and well-documented. Thanks!

avisionh committed:
docs: Add title and axis labels

This is so the plot is publication ready.

avisionh committed:
feat: Add factor analysis

This is to understand the latent relationships between the variables.

avisionh committed:
feat: Add VIP for random forest

This is so we can check the important variables are important across the two models, acting as a QA check.

avisionh committed:
feat: Add checks on vip

This is to see if the identified variables in vip have a relatively high association with wellbeing.

avisionh commented:
Thanks for the review, @martinctc! I've implemented the changes you suggested though was a bit unsure why the factor analysis led to 0 factors being recommended. 

> Great work @avisionh!
> 
> I've reviewed this PR and had the following observations:
> 
> 1. Computing lag for IV is a nice detail to add - I like this!
> 2. Would it be worth running PCA or factor analysis on the predictors, if you haven't yet? Since multi-collinearity is a typical challenge with survey datasets.
- Good idea, done in https://github.com/ukgovdatascience/surveyanalysis/pull/5/commits/1e3b92d7f0fc0c5fa79c45bcccf6140ab3da6ee6.
> 3. Agree with multinomial regression and RF as the model choices to try with this set of data.
> 4. When you run `vip()` for either models, do the models come up with similar results on the most important independent variables? It's usually difficult to get good accuracy on survey data, so I'd typically also do a sense check on whether the results make sense, based on domain knowledge / context.
- Yup, they do, but I made it more explicit here. https://github.com/ukgovdatascience/surveyanalysis/pull/5/commits/afc061fb8051bc61b618af1d67482c9d271f62e4
> 5. Also, how do the variable importance plots compare with correlation coefficients vs the dependent variable? It's usually helpful to be able to explain in _both_ cases of (i) either they are similar or (ii) very different.
- Nice suggestion, fits in domain knowledge too. https://github.com/ukgovdatascience/surveyanalysis/pull/5/commits/b0b72483f0e17d7cb219c32f109cfed5b903e473
> 6. Presumably the data covers all the pupils in the analysis 'universe'? i.e. it's technically a census rather than a sample of the universe? If it's the latter, it could be worth thinking about weighting, but this might not be necessary if we think it's a fair representation of the population.
- I think it's the best we can do. Not sure if we know what the population weights would be here but it's a good thing to add in the caveats section of the report!
> 
> Note that I did not review the code line-by-line, and the notes are only based on a review of the overall methodology. The code however looks very neat and well-documented. Thanks!
